README.md
# Html5ZipFile

HTML 5 zip file validation and unpacking.

Project home: https://github.com/adgear/html5_zip_file

Yard docs: https://adgear.github.io/html5_zip_file/frames.html

Travis CI: https://travis-ci.org/adgear/html5_zip_file

[![Build Status](https://travis-ci.org/adgear/html5_zip_file.svg?branch=ruby-1-8-7-compat)](https://travis-ci.org/adgear/html5_zip_file)

## Install for development

Perform development with the current version of ruby (2.2.2).

    git clone git@github.com:adgear/html5_zip_file.git

Install dependencies:

    $ gem install bundler
    $ bundle install

Run tests:

    $ bundle exec rake test
    $ bundle exec ruby -Itest test/file_test.rb -n "/.*validate_valid_zip.*/"

## Usage

Run doctests: {HTML5ZipFile::File.open},
{HTML5ZipFile::File#validate} and {HTML5ZipFile::File#unpack}.

    $ bundle exec yard doctest -v

    # Running:

    entries:
    #<ZipUnpack::Entry:0x007ff943af05c0 @ftype=:file, @name="index.html", @size=112>
    #<ZipUnpack::Entry:0x007ff943af0480 @ftype=:directory, @name="images/", @size=0>
    #<ZipUnpack::Entry:0x007ff943af02c8 @ftype=:file, @name="images/test.png", @size=732059>
    #<ZipUnpack::Entry:0x007ff943af0138 @ftype=:directory, @name="foo/", @size=0>
    #<ZipUnpack::Entry:0x007ff943aebf98 @ftype=:file, @name="foo/index.html", @size=62>
    #<ZipUnpack::Entry:0x007ff943aebea8 @ftype=:file, @name="foo/index2.html", @size=41>
    HTML5ZipFile::File.open#test_0001_Open a zip file = 0.01 s = .

    Failed validation checks:
    size_unpacked
    file_entries
    HTML5ZipFile::File#validate#test_0001_Validate a zip file = 0.01 s = .

    /var/folders/_h/d4trcrkn3mv3w307j7_w3xbh0000gn/T/HTML5ZipFile_extract_20160115-67900-hdxmoc
    .
    ..
    foo
    images
    index.html
    HTML5ZipFile::File#unpack#test_0001_Unpack a zip file = 0.02 s = .

    Finished in 0.055291s, 54.2580 runs/s, 54.2580 assertions/s.

    3 runs, 3 assertions, 0 failures, 0 errors, 0 skips

## Integrate

Add the appropriate line to your application's Gemfile:

    gem 'html5_zip_file', :git => 'git://github.com/adgear/html5_zip_file'

    gem 'html5_zip_file_1_8_7', :git => 'git://github.com/adgear/html5_zip_file'

Execute:

    $ bundle install

See {file:test/kitchen_sink.rb kitchen_sink.rb} for an example of how to actually use the code.

    $ bundle exec ruby test/kitchen_sink.rb
    I, [2016-01-15T11:15:18.107693 #68031]  INFO -- : Info-ZIP: found version UnZip 5.52
    I, [2016-01-15T11:15:18.114819 #68031]  INFO -- : Info-ZIP: CRC check passed
    I, [2016-01-15T11:15:18.116761 #68031]  INFO -- : Info-ZIP: entries parsed

    size_unpacked: 732274

    #<ZipUnpack::Entry:0x007fa89a161170 @ftype=:file, @name="index.html", @size=112>
    #<ZipUnpack::Entry:0x007fa89a161030 @ftype=:directory, @name="images/", @size=0>
    #<ZipUnpack::Entry:0x007fa89a160ec8 @ftype=:file, @name="images/test.png", @size=732059>
    #<ZipUnpack::Entry:0x007fa89a160db0 @ftype=:directory, @name="foo/", @size=0>
    #<ZipUnpack::Entry:0x007fa89a160c98 @ftype=:file, @name="foo/index.html", @size=62>
    #<ZipUnpack::Entry:0x007fa89a160ba8 @ftype=:file, @name="foo/index2.html", @size=41>

    Failed validation checks:
    size_unpacked
    file_entries
    path_length
    contains_html_file

## Docs / Doctests

Use current ruby.

http://yardoc.org

    $ bundle exec yard help

    $ bundle exec yard doc --files test/kitchen_sink.rb

    $ bundle exec yard server -r
    $ rm -rf .yardoc/
    $ bundle exec yard server -r -m
    $ bundle exec yard help server

    $ bundle exec yard list --query '@todo'

    $ bundle exec yard graph --full --dependencies | dot -Tpng -o outfile.png

https://github.com/p0deje/yard-doctest

    $ bundle exec yard config load_plugins true

    $ bundle exec yard config -a autoload_plugins yard-doctest

    $ bundle exec yard doctest -h

    $ bundle exec yard doctest -v

## External Dependency

This gem has one external dependency: the command line program 'unzip'.

http://www.info-zip.org/UnZip.html

The test suite guarantees compabitibility with unzip v5.52 & v6.0.

Versions currently provided by various distributions (2015-12-21):

- OSX: 5.52 (2005)
- Arch: 6.0-11 (4 fixes in 2015)
- Gentoo: 6.0-r3 (1 change in 2015)
- Debian: 6.0-16 (4 fixes in 2015)
- Centos: 6.0-15

If you use html5_zip_file in an application, you should run the test
suite on a production machine, and "freeze" the unzip version in the
OS's package manager.

Any time the package manager wants to update unzip, "rake test" should
be run and the gem updated as necessary.

## Security

A first layer of security is provided by imposing sensible limits on
:size_unpacked, :entries, :path_length and :path_components via
{HTML5ZipFile::File#validate}.

### Zip entry metadata

Validation requires running the "unzip -t" and "unzip -l" commands
in subprocesses, and parsing the stdout in a ruby parent process.

### Zip unpacking

Currently, the file is unpacked to a target directory with 'unzip -d'.

We rely on unzip's internal checks to prevent unpacking of files to
locations outside the target directory.

Stronger guarantees could be provided by unpacking in a sandbox:

- Create a temporary sandbox directory with Ruby's tmpdir module.

- Rename the zip file and copy into the sandbox.

- Child process (without the ability to write to filesystem locations
  outside the sandbox) runs 'unzip -d' to a subdirectory in the
  sandbox.

- Once the child terminates, the parent copies the extracted files
  from the sandbox to their final location.

- Parent destroys the sandbox.

Steps to increase the security of the sandbox:

1.) Mount the temporary filesystem on a separate partition.

Determine the location tmpdir uses through irb:

Ruby 2.2.2 on OSX:

    $ irb
    > require 'tmpdir'
    > Dir.mktmpdir do |d|
    puts d.inspect()
    > end
    "/var/folders/_h/d4trcrkn3mv3w307j7_w3xbh0000gn/T/d20151223-49143-1mxf2k4"

Ruby 2.2.2 on debian: "/tmp/d20151223-28897-7wfgjm"

2.) Use linux namespaces to prevent filesystem access outside the
sandbox by the subprocess.

3.) Drop unneeded capabilities and put resource utilization limits on
the subprocess.

## Gems

Separate gemspecs are maintained for 'current' ruby and ruby 1.8.7.

Current ruby has rubygems built-in.

    $ chruby 2.2.2

    $ gem build html5_zip_file.gemspec
    Successfully built RubyGem
    Name: html5_zip_file
    Version: 1.0
    File: html5_zip_file-1.0.gem

    $ gem install [--dev] html5_zip_file-1.0.gem
    Successfully installed html5_zip_file-1.0
    Parsing documentation for html5_zip_file-1.0
    Installing ri documentation for html5_zip_file-1.0
    Done installing documentation for html5_zip_file after 0 seconds

    $ gem which html5_zip_file
    /Users/MYUSER/.gem/ruby/2.2.2/gems/html5_zip_file-1.0/lib/html5_zip_file.rb

    $ irb
    irb(main):001:0> require 'html5_zip_file'

Ruby 1.8.7 does not have rubygems built-in: install rubygems 1.3.6
manually.

    $ chruby 1.8.7

    $ gem --version
    1.3.6

    $ gem build html5_zip_file_1_8_7.gemspec
    Successfully built RubyGem
    Name: html5_zip_file_1_8_7
    Version: 1.0
    File: html5_zip_file_1_8_7-1.0.gem

    $ gem install [--dev] html5_zip_file_1_8_7-1.0.gem
    Successfully installed html5_zip_file_1_8_7-1.0
    1 gem installed
    Installing ri documentation for html5_zip_file_1_8_7-1.0...
    Installing RDoc documentation for html5_zip_file_1_8_7-1.0...

    $ gem which html5_zip_file
    /Users/MYUSER/.gem/ruby/1.8.7/gems/html5_zip_file_1_8_7-1.0/lib/html5_zip_file.rb

    $ irb
    irb(main):001:0> require 'rubygems'
    irb(main):002:0> require 'html5_zip_file'
    irb(main):003:0> HTML5ZipFile::File
    => HTML5ZipFile::File

In theory, you could build gems, push them to a repository such as
rubygems.org, and install them in your destination environment with
"gem install XXX".

## License

The gem is available as open source under the terms of the [MIT
License](http://opensource.org/licenses/MIT).


RELEASE_NOTES
Version 0.4.0 [Released January 19 2016]

* Refocus on core functionality
  * Remove inject_script_tag()
  * Remove :contains_zip_file
  * Remove destroy_unpacked()

* Make compat with ruby 1.8.7
  * Modify tooling 
    * Remove superfluous bin scripts
    * Remove superfluous version file
    * Remove $LOAD_PATH mangling
    * Remove bundler rake tasks
    * Remove minitest reporters
    * Remove byebug, pry and bundler
    * Make rubygems work under ruby 1.8.7
    * Add gemspec for ruby 1.8.7
  * Refactor HTML5ZipFile::File (remove rubyzip & improve interface)
    * Replace rubyzip with posix-spawn / 'unzip'
    * Add :size_packed
    * Rename :contents_size to :size_unpacked
    * Rename :entry_count to :entries
    * Rename :file_count to :file_entries
    * Rename :directory_count to :directory_entries
    * Remove :contains_zip_file
    * Hide constructor
    * Improve exceptions

* Add doctests

* Add kitchen sink example

* Set up CI


test/file_test.rb
require 'test_helper'
require 'html5_zip_file/file'
require 'byebug'

module HTML5ZipFile
  class FileTest < Minitest::Test
    def test_validate_valid_zip
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate
      end
    end

    def test_validate_invalid_zip
      File.open('test/data/invalid.zip') do |f|
        refute f.validate
        assert_equal [:zip], f.failures
      end
    end

    def test_validate_valid_contents__size
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:contents_size => 20_000_000)
      end
    end

    def test_validate_invalid_contents_size
      File.open('test/data/test-ad.zip') do |f|
        refute f.validate(:contents_size => 100_000)
        assert_equal [:contents_size], f.failures
      end
    end

    def test_validate_valid_entry_count
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:entry_count => 6)
      end
    end

    def test_validate_invalid_entry_count
      File.open('test/data/test-ad.zip') do |f|
        refute f.validate(:entry_count => 5)
        assert_equal [:entry_count], f.failures
      end
    end

    def test_validate_valid_file_count
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:file_count => 4)
      end
    end

    def test_validate_invalid_file_count
      File.open('test/data/test-ad.zip') do |f|
        refute f.validate(:file_count => 3)
        assert_equal [:file_count], f.failures
      end
    end

    def test_validate_valid_directory_count
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:directory_count => 2)
      end
    end

    def test_validate_invalid_directory_count
      File.open('test/data/test-ad.zip') do |f|
        refute f.validate(:directory_count => 1)
        assert_equal [:directory_count], f.failures
      end
    end

    def test_validate_valid_path_length
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:path_length => 15)
      end
    end

    def test_validate_invalid_path_length
      File.open('test/data/test-ad.zip') do |f|
        refute f.validate(:path_length => 14)
        assert_equal [:path_length], f.failures
      end
    end

    def test_validate_valid_path_components
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:path_components => 2)
      end
    end

    def test_validate_invalid_path_components
      File.open('test/data/test-ad.zip') do |f|
        refute f.validate(:path_components => 1)
        assert_equal [:path_components], f.failures
      end
    end

    def test_validate_valid_contains_html_file
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:contains_html_file => true)
      end
    end

    def test_validate_invalid_contains_html_file
      File.open('test/data/test-ad-no-html.zip') do |f|
        refute f.validate(:contains_html_file => true)
        assert_equal [:contains_html_file], f.failures
      end
    end

    def test_validate_valid_contains_zip_file
      File.open('test/data/test-ad.zip') do |f|
        assert f.validate(:contains_zip_file => false)
      end
    end

    def test_validate_invalid_contains_zip_file
      File.open('test/data/test-ad-with-zip.zip') do |f|
        refute f.validate(:contains_zip_file => false)
        assert_equal [:contains_zip_file], f.failures
      end
    end

    def test_validate_mixed
      File.open('test/data/test-ad.zip') do |f|
        valid = f.validate(
          :size => 1_000_000,
          :entry_count => 6,
          :file_count => 2,
          :directory_count => 2,
          :path_length => 10,
          :path_components => 5,
          :contains_html_file => true,
          :contains_zip_file => false
        )
        refute valid
        assert_equal 2, f.failures.size
        assert_includes f.failures, :file_count
        assert_includes f.failures, :path_length
      end
    end

    def test_validate_invalid_zip_mixed
      File.open('test/data/invalid.zip') do |f|
        refute f.validate(:entry_count => 6, :file_count => 2)
        assert_equal [:zip], f.failures
      end
    end

    def test_validate_forbidden_characters
      File.open('test/data/test-ad-with-spaces.zip') do |f|
        refute f.validate(:forbidden_characters => /[ ]/)
        assert_equal [:forbidden_characters], f.failures
      end
    end

    def test_contents_size
      File.open('test/data/test-ad.zip') do |f|
        assert_equal 732274, f.contents_size
      end
    end

    def test_entries
      File.open('test/data/test-ad.zip') do |f|
        assert_equal 6, f.entries.size
      end
    end

    def test_file_entries
      File.open('test/data/test-ad.zip') do |f|
        assert_equal 4, f.file_entries.size
      end
    end

    def test_directory_entries
      File.open('test/data/test-ad.zip') do |f|
        assert_equal 2, f.directory_entries.size
      end
    end

    def test_html_file_entries
      File.open('test/data/test-ad.zip') do |f|
        assert_equal 3, f.html_file_entries.size
      end
    end

    class Unpack < Minitest::Test
      def teardown
        FileUtils.rm_rf('test/unpack')
      end

      def test_unpack_non_empty
        assert_raises(DestinationNotEmpty) do
          File.open('test/data/test-ad.zip') do |f|
            f.unpack('test/unpack_non_empty')
          end
        end
      end

      def test_unpack_empty
        FileUtils.mkdir_p('test/unpack')

        File.open('test/data/test-ad.zip') do |f|
          f.unpack('test/unpack')

          entries = Dir.entries('test/unpack')
          assert_equal 5, entries.size
          assert_includes entries, 'index.html'
          assert_includes entries, 'images'
          assert_includes entries, 'foo'

          entries = Dir.entries('test/unpack/images')
          assert_equal 3, entries.size
          assert_includes entries, 'test.png'

          entries = Dir.entries('test/unpack/foo')
          assert_equal 4, entries.size
          assert_includes entries, 'index.html'
          assert_includes entries, 'index2.html'
        end
      end

      def test_unpack_mkdir
        File.open('test/data/test-ad.zip') do |f|
          f.unpack('test/unpack')

          entries = Dir.entries('test/unpack')
          assert_equal 5, entries.size
          assert_includes entries, 'index.html'
          assert_includes entries, 'images'
          assert_includes entries, 'foo'

          entries = Dir.entries('test/unpack/images')
          assert_equal 3, entries.size
          assert_includes entries, 'test.png'

          entries = Dir.entries('test/unpack/foo')
          assert_equal 4, entries.size
          assert_includes entries, 'index.html'
          assert_includes entries, 'index2.html'
        end
      end

      def test_destroy_unpacked
        File.open('test/data/test-ad.zip') do |f|
          f.unpack('test/unpack')
          f.destroy_unpacked

          refute Dir.exists?('test/unpack')
        end
      end

      def test_inject_script_tag_not_unpacked
        assert_raises(NotUnpacked) do
          File.open('test/data/test-ad.zip') do |f|
            f.inject_script_tag('<script></script>')
          end
        end
      end

      def test_inject_script_tag_invalid
        assert_raises(InvalidScriptTag) do
          File.open('test/data/test-ad.zip') do |f|
            f.unpack('test/unpack')
            f.inject_script_tag('<div></div>')
          end
        end
      end

      def test_inject_script_tag
        File.open('test/data/test-ad.zip') do |f|
          f.unpack('test/unpack')
          f.inject_script_tag('<script src="test.js"></script>')

          data = ::File.read('test/unpack/index.html')
          assert data.start_with?('<!DOCTYPE html>')
          html = Nokogiri::HTML.parse(data)
          assert html.at_css('head script[src="test.js"]')

          data = ::File.read('test/unpack/foo/index.html')
          assert data.start_with?('<!DOCTYPE html>')
          html = Nokogiri::HTML.parse(data)
          assert html.at_css('html script[src="test.js"]')

          data = ::File.read('test/unpack/foo/index2.html')
          assert data.start_with?('<!DOCTYPE html>')
          html = Nokogiri::HTML.parse(data)
          assert html.at_css('script[src="test.js"]')
        end
      end

      def test_inject_script_tag_replace
        File.open('test/data/test-ad.zip') do |f|
          f.unpack('test/unpack')
          f.inject_script_tag('<script src="a.js"></script>')
          f.inject_script_tag('<script src="b.js"></script>')

          data = ::File.read('test/unpack/index.html')
          html = Nokogiri::HTML.parse(data)

          assert html.at_css('script[src="b.js"]')
          refute html.at_css('script[src="a.js"]')
        end
      end
    end
  end
end


test/test_helper.rb
require 'rubygems'

require 'minitest/autorun'

require 'html5_zip_file'


test/kitchen_sink.rb
require 'html5_zip_file'
require 'tmpdir'

ZipUnpack::ZipFile.set_log_level Logger::INFO

HTML5ZipFile::File.open('test/data/test-ad.zip') do |f|

  puts "size_unpacked: #{f.size_unpacked}"

  f.entries.each { |e| puts(e.inspect()) }

  if f.validate( :size_unpacked => 700000,
                 :file_entries => 3,
                 :path_length => 14,
                 :contains_html_file => false )

    Dir.mktmpdir("HTML5ZipFile_extract_") do |dir|
      f.unpack(dir)
    end

  else

    puts 'Failed validation checks: '
    f.failures.each { |failure|  puts(failure) }

  end

end


lib/html5_zip_file/file.rb
require 'fileutils'
require 'nokogiri'
require 'pathname'
require 'zip'

module HTML5ZipFile
  class File
    # Array of validation failure keywords.
    attr_reader :failures

    # Open a zip file by path.
    def self.open(file_name)
      Zip::File.open(file_name, 0) do |z|
        yield File.new(z)
      end
    rescue Zip::Error
      yield File.new(nil)
    end

    def initialize(zip) # :nodoc:
      @zip = zip
    end

    # Validate zip file with options:
    #
    # - +:contents_size+:: maximum contents size
    # - +:entry_count+:: maximum number of entries (files and directories)
    # - +:file_count+:: maximum number of file
    # - +:directory_count+:: maximum number of directories
    # - +:path_length+:: maximum path length in characters
    # - +:path_components+:: maximum number of path components
    # - +:contains_html_file+:: require or disallow HTML files to exist
    # - +:contains_zip_file+:: require or disallow embedded zip files to exist
    # - +:forbidden_characters:: disallow characters that match this regexp in any entry
    #
    # Populates #failures with keywords of failed validations. If file is not
    # a valid zip, #failures will contain +:zip+.
    #
    # Returns +true+ if validations passed.
    def validate(options = {})
      @failures = []

      unless is_valid_zip?
        @failures << :zip
        return false
      end

      if options.has_key? :contents_size
        @failures << :contents_size unless is_valid_contents_size? options[:contents_size]
      end

      if options.has_key? :entry_count
        @failures << :entry_count unless is_valid_entry_count? options[:entry_count]
      end

      if options.has_key? :file_count
        @failures << :file_count unless is_valid_file_count? options[:file_count]
      end

      if options.has_key? :directory_count
        @failures << :directory_count unless is_valid_directory_count? options[:directory_count]
      end

      if options.has_key? :path_length
        @failures << :path_length unless is_valid_path_length? options[:path_length]
      end

      if options.has_key? :path_components
        @failures << :path_components unless is_valid_path_components? options[:path_components]
      end

      if options.has_key? :contains_html_file
        @failures << :contains_html_file unless contains_html_file? == options[:contains_html_file]
      end

      if options.has_key? :contains_zip_file
        @failures << :contains_zip_file unless contains_zip_file? == options[:contains_zip_file]
      end

      if options.has_key? :forbidden_characters
        @failures << :forbidden_characters if
          entries.map(&:name).any? { |n| n.match(options[:forbidden_characters]) }
      end

      @failures.none?
    end

    # Return total size of zip file contents.
    def contents_size
      @contents_size ||= file_entries.map(&:size).reduce(0, :+)
    end

    # Return all entries.
    def entries
      @zip.entries
    end

    # Return file entries.
    def file_entries
      @file_entries ||= @zip.select do |entry|
        entry.ftype == :file
      end
    end

    # Return directory entries.
    def directory_entries
      @directory_entries ||= @zip.select do |entry|
        entry.ftype == :directory
      end
    end

    # Return HTML file entries.
    def html_file_entries
      @html_file_entries ||= file_entries.select do |entry|
        entry.name =~ /\.html?\z/i
      end
    end

    # Unpack the zip file to +dest+, which must either be an empty directory
    # or not exist.
    #
    # If the destination is not empty, raises #DestinationNotEmpty.
    def unpack(dest)
      exists = Dir.exists?(dest)
      if exists && Dir.entries(dest).size > 2
        raise DestinationNotEmpty, 'Destination directory is not empty'
      end
      FileUtils.mkdir_p(dest) unless exists

      @unpack_dest = dest

      @zip.each do |entry|
        entry.extract(::File.join(dest, entry.name))
      end
    end

    # Destroy previously unpacked contents.
    def destroy_unpacked
      return unless @unpack_dest && Dir.exists?(@unpack_dest)
      FileUtils.remove_entry_secure(@unpack_dest, :force => true)
      @unpack_dest = nil
    end

    # Inject a script tag into alll unpacked HTML files, replacing any
    # previously injected script tag.
    #
    # Raises #NotUnpacked if #unpack has not been called.
    # Raises #InvalidScriptTag if +script_tag+ does not contain a valid tag.
    def inject_script_tag(script_tag)
      raise NotUnpacked, 'Zip file not unpacked' unless @unpack_dest

      html_file_entries.each do |entry|
        path = ::File.join(@unpack_dest, entry.name)

        data = ::File.read(path)
        html = Nokogiri::HTML.parse("<!DOCTYPE html>\n" + data)

        script_fragment = Nokogiri::HTML.fragment(script_tag)
        tag = script_fragment.at_css('script')
        raise InvalidScriptTag, 'Script tag missing' unless tag
        tag['data-adgear-html5'] = 'true'

        if existing_tag = html.at_css('script[data-adgear-html5="true"]')
          existing_tag.replace(tag)
        elsif head_tag = html.at_css('head')
          head_tag.prepend_child(tag)
        elsif html_tag = html.at_css('html')
          html_tag.prepend_child(tag)
        else
          html.prepend_child(tag)
        end

        ::File.write(path, html.to_s)
      end
    end

    private

    def is_valid_zip?
      !!@zip
    end

    def is_valid_contents_size?(max)
      contents_size <= max
    end

    def is_valid_entry_count?(max)
      @zip.size <= max
    end

    def is_valid_file_count?(max)
      file_entries.size <= max
    end

    def is_valid_directory_count?(max)
      directory_entries.size <= max
    end

    def is_valid_path_length?(max)
      @zip.all? do |entry|
        entry.name.size <= max
      end
    end

    def is_valid_path_components?(max)
      @zip.all? do |entry|
        Pathname.new(entry.name).each_filename.count <= max
      end
    end

    def contains_html_file?
      html_file_entries.any?
    end

    def contains_zip_file?
      zip_file_entries.any?
    end

    def zip_file_entries
      @zip_file_entries ||= file_entries.select do |entry|
        entry.name =~ /\.zip\z/i
      end
    end
  end
end


lib/html5_zip_file/zip_unpack.rb
require 'logger'

require 'rubygems'
require 'posix/spawn'

# https://www.pkware.com/support/zip-app-note/
# https://en.wikipedia.org/wiki/Zip_(file_format)
# https://users.cs.jmu.edu/buchhofp/forensics/formats/pkzip.html

module ZipUnpack

  Error = Class.new(StandardError)

  CorruptZipFileError = Class.new(Error)

  class Entry
    include Comparable
    attr_reader :ftype, :name, :size
    def initialize(ftype, name, size)
      @ftype = ftype
      @name = name
      @size = size
    end
    def <=>(another)
      name <=> another.name
    end
  end

  # Methods that use POSIX::Spawn to run external processes
  # ({#initialize} and {#unpack}) may raise the following exceptions:
  #
  # @raise [POSIX::Spawn::MaximumOutputExceeded] if output length > SPAWN_CHILD_OPTS:max bytes
  # @raise [POSIX::Spawn::TimeoutExceeded] execution time > SPAWN_CHILD_OPTS:timeout seconds
  class ZipFile

    # Compatible Info-ZIP versions
    VERSION_WHITELIST = ['UnZip 5.52', 'UnZip 6.0']

    # Security limits for POSIX::Spawn::Child
    SPAWN_CHILD_OPTS = {
      :max => 300_000,
      :timeout => 20
    }

    ZipFileError = Class.new(StandardError)
    UnzipBadVersionError = Class.new(ZipFileError)
    UnzipParsingError = Class.new(ZipFileError)

    @@log = Logger.new(STDOUT)
    @@log.level = Logger::WARN

    def self.set_log_level(level)
      @@log.level = level
    end

    # @raise [Errno::ENOENT] unzip command line binary not found in $PATH
    # @raise [UnzipBadVersionError] unzip version string not whitelisted
    # @raise [UnzipParsingError] unable to parse output from unzip command
    #
    # @raise [CorruptZipFileError] the zip file is corrupt
    def initialize(filename)
      @filename = filename

      # Check unzip command presence & version
      ver = get_infozip_version
      @@log.info "Info-ZIP: found version #{ver}"

      # Check zip file integrity
      raise CorruptZipFileError unless crc_valid?(filename)
      @@log.info 'Info-ZIP: CRC check passed'

      # Get zip file entries
      @entries = get_entries(filename)
      @@log.info 'Info-ZIP: entries parsed'
    end

    def unpack(dest)
      @@log.info "Info-ZIP: unpacking to #{dest}"
      child = POSIX::Spawn::Child.new('unzip', '-d', dest, @filename, SPAWN_CHILD_OPTS)
      raise CorruptZipFileError, "Unzip failed" if child.status.exitstatus != 0
      @@log.info "Info-ZIP: unpack succeeded"
    end

    attr_reader :entries

    def size_packed
      @packed_size ||= File.size(@filename)
    end

    private

    # @return [String] version_string whitelisted version_string found
    def get_infozip_version
      child = POSIX::Spawn::Child.new('unzip', '-v', SPAWN_CHILD_OPTS)
      parse_infozip_version_string(child.out)
    end

    # @return [String] version_string whitelisted version_string found
    def parse_infozip_version_string(stdout)
      VERSION_WHITELIST.each do |ver|
        return ver if Regexp.new('\A' + ver) =~ stdout
      end
      raise UnzipBadVersionError
    end

    # Extract the files in memory and compare the CRC of each
    # expanded file with the original file's stored CRC value.
    #
    # @TODO: test memory/cpu behavior on zip bombs
    def crc_valid?(filename)
      child = POSIX::Spawn::Child.new('unzip', '-t', filename, SPAWN_CHILD_OPTS)
      child.status.exitstatus == 0
    end

    # Get zip file entries
    #
    # @return [Array<Entry>] array of entries
    def get_entries(filename)
      child = POSIX::Spawn::Child.new('unzip', '-l', filename, SPAWN_CHILD_OPTS)
      raise CorruptZipFileError if child.status.exitstatus != 0
      parse_entries(child.out)
    end

    #@TODO: test memory/cpu behavior on zip bombs
    def parse_entries(stdout)
      lines = stdout.split("\n")
      raise UnzipParsingError, 'unzip -l output has too few lines.' \
                             if lines.size < 5
      entry_lines = lines.slice(3, lines.count - 5)
      entries = []
      entry_lines.each do |entry_line|
        if /^\s*(\d+)\s+\d+-\d+-\d+\s+\d+:\d+\s+(.+)$/ =~ entry_line
          entry_size = $1.to_i
          entry_name = $2
          if entry_name[-1, 1] == '/'
            raise UnzipParsingError, "Directory entry with non-zero size." if entry_size != 0
            entries << Entry.new(:directory, entry_name, entry_size)
          else
            entries << Entry.new(:file, entry_name, entry_size)
          end
        else
          raise UnzipParsingError, 'Bad entry line.'
        end
      end
      entries
    end
  end

  class CorruptZipFile
    def unpack(d)
      raise CorruptZipFileError
    end
    def size_packed
      raise CorruptZipFileError
    end
    def entries
      raise CorruptZipFileError
    end
  end

end
